### PR TITLE
fix(sql): breaking change💥 - fix expression parser operator precedence

### PR DESCRIPTION
**Pull request by sivukhin**
## Context

This is second part of the PR https://github.com/questdb/questdb/pull/4386

This PR fixes expression parser operator precedence rules. Now, `QuestDB` will use precedence order which is close to the PostgreSQL  (https://www.postgresql.org/docs/7.2/sql-precedence.html; the precedence rules validated extensively with ExpressionParserFuzzTest.java):
```
{"."},
{"::"},
{"-'1", "~'1"},
{"*", "/", "%"},
{"+", "-'2"},
{"<<", ">>", "<<=", ">>="},
{"||"},
{"&"},
{"^"},
{"|"},
{"in", "not in", "within", "not within", "between", "not between"},
{"<", "<=", ">", ">="},
{"=", "~'2", "!=", "<>", "!~", "like", "ilike", "is null", "is not null"},
{"not"},
{"and"},
{"or"}
```

In order to provide migration path for users `QuestDB` introduces **temporary** boolean flag in the config: `cairo.sql.legacy.operator.precedence` or `QDB_CAIRO_SQL_LEGACY_OPERATOR_PRECEDENCE` env var which enables legacy operator precedence in a switched-on (`true`) state + with `cairo.sql.legacy.operator.precedence = true` QuestDB will log mismatch in parsing behaviour between old & new precedence table:
```
2024-04-25T21:23:19.408159Z A i.q.g.ExpressionParser operator precedence compat mode: detected expression parsing behaviour change for query=[true or true and false] at position=13
```

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/ExpressionParser.java
Comment (sivukhin) [reply core/src/main/java/io/questdb/griffin/ExpressionParser.java:1364]: > You can paste example queries as docstrings, the kind of queries you'd run in tests

My first todo comment were about log entry: currently I log only expression part (so `WHERE` clause for example) - but user will not see full query in the log (which can have table name for example - and this can help user to identify source of these queries).

It's a bit non-trivial thing to do - as full query parsing is even more complicated part compared to `ExpressionParser`and I'm afraid that a lot of code will be touched in order to show full query instead of broken clause...

File: core/src/main/java/io/questdb/PropServerConfiguration.java
Comment (sivukhin) [reply core/src/main/java/io/questdb/PropServerConfiguration.java]: Replaced this string parameter with boolean flag and removed version from its name

File: core/src/main/java/io/questdb/griffin/ExpressionParser.java
Comment (sivukhin) [reply core/src/main/java/io/questdb/griffin/ExpressionParser.java:1364]: Yes, I understand.

> Also, we can add link to the doc/migration guide

Regarding this - I agreed this is not necessary.

> Can we add full query here? Or maybe some trace id?

But here I don't know what to do. You suggest just to leave logging as is, right? There is no way to log some trace id maybe (do `QuestDB` have something like that) or get access to the full query instead of only expression clause?

Comment (sivukhin) [reply core/src/main/java/io/questdb/griffin/ExpressionParser.java:1364]: > Can a user be confident that everything is fine if this log never appears?

Actually, this can be done. Current version in the PR handles almost all cases except `NOT` operator. But I guess we can make this work too - and then it will be the guarantee: if no warning of this format appeared in the logs - then there were no queries with "ambiguous" parsing in the system (for sure, there can be a query which will happen once in a month - but it's out of scope of this guarantee). 

> Why should they rely on a log message (which might be disabled in production) for safety?

I'm actually think that this is fine to some extent. Me, as an external user of several databases, look at the logs as to an additional database "instrument". It has very vague API - but I can use it (with `grep` for example) and analyze to some extent. So I think this is fine to use logs in some case of breaking changes (like this). Plus, I guess `QuestDB` already report some warnings in the log (for example, deprecated parameters) - don't see any difference with our situation here.

> Imo, its false security and users should review the situation themselves and not rely on a log message to determine whether they have a problem or not

I think I disagree here too. It seems to me that with this kind of arguments everything can be called as "false security" and finally you end up forcing users to read all patches to `QuestDB` / `JVM` / `Linux kernel` / etc. 

Also, as I sad it earlier, from the user perspective I find it very unsatisfying to have migration plan which includes step "review all your queries to not have expression with ambiguous binary operator parsing". If I were maintainer of `QuestDB` installation - I may even not know "all sources" of queries and in medium/large-size org this migration can spans months or even more (this obviously based on some my experience - not `QuestDB` - just general SWE experience). And this is kind of tedious work to find all usages, validate them manually and fix them. 

So, I think it's really important to provide more smooth migration path which will help users to adopt new version more easily. Sure, there are users who disable logs for production cluster - but they should understand that this can hurts them at some point (for example, if something will went wrong with their `QuestDB` instances - there will be only limited support for their issue from the development/support team and this is totally fine; if they disable logs, `QuestDB` should not stop writing them at all - because many users still have logs enabled for their production clusters I guess).

---

Definitely, more opinions will be helpful here (I feel that we are going in circles and can't convince each other - so need more people to have quorum over final decision). 

Not sure about second reviewer - maybe it's better to bring this discussion in "more public" (compared to the PR comments) space independently from review? @nwoolmer, what do you think?

File: core/src/main/java/io/questdb/PropServerConfiguration.java
Comment (sivukhin) [reply core/src/main/java/io/questdb/PropServerConfiguration.java:1103]: Hm...yes, `true` = new precedence, you are right. But I don't want to introduce breaking change in this PR - so want to make old behaviour default one...

Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/PropServerConfiguration.java:1121]: I agree with Nick: the default should be breaking behavior. Also, it may be a good idea to name the prop something like `cairo.sql.legacy.operator.precedence`.

File: core/src/test/java/io/questdb/test/griffin/ExpressionParserFuzzTest.java
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/test/griffin/ExpressionParserFuzzTest.java:1]: I guess the commented part from `ExpressionParserTest#testCorrectPrecedenceOfBasicOps()` should be uncommented by this PR.

**Comment by puzpuzpuz** [comment]
> In order to provide migration path for users QuestDB introduces temporary flag in the config

This part of the PR description seems no longer relevant and should be updated.

File: core/src/main/java/io/questdb/griffin/OperatorRegistry.java
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/OperatorRegistry.java:41]: nit: these local fields are not required. Instead, class fields can be assigned in the below loop and later asserted.

File: core/src/main/java/io/questdb/griffin/OperatorRegistry.java
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/OperatorRegistry.java:108]: nit:
```suggestion
        return map.contains(name);
```

File: core/src/main/java/io/questdb/griffin/OperatorRegistry.java
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/OperatorRegistry.java:82]: This method, as well as `tryGuessOperator`, deserve a javadoc clarifying their purpose.

File: core/src/main/java/io/questdb/griffin/SqlParser.java
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlParser.java:27]: This import seems to be unused - the same applies to other classes changed by the PR. If you're using IDEA, please run the Optimize imports action.

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:5019]: Shouldn't we use new or legacy registry here and below based on the settings?

File: core/src/main/java/io/questdb/griffin/SqlOptimiser.java
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlOptimiser.java:3517]: Again, we should probably select the registry based on the settings.

File: core/src/main/java/io/questdb/griffin/SqlParser.java
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/griffin/SqlParser.java:109]: By default, we should be using the new registry as the active one and the legacy registry as the shadow one, but I don't see this here and in the below branch. Am I missing something?

File: core/src/test/java/io/questdb/test/griffin/ExpressionParserFuzzTest.java
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/test/griffin/ExpressionParserFuzzTest.java:42]: Do you need this test to be `AbstractBootstrapTest`? Those tests start all configured servers, so they're relatively slow. In most cases, `AbstractCairoTest` should be more than enough since it only runs the engine.

File: core/src/main/java/io/questdb/griffin/SqlParser.java
Comment (sivukhin) [reply core/src/main/java/io/questdb/griffin/SqlParser.java]: Tried to optimize all imports. But Optimize imports action over `src/` dir yield a lot changes in other files for me - so I rolled them back and leaved only files changes in my PR

File: core/src/main/java/io/questdb/PropServerConfiguration.java
Comment (sivukhin) [reply core/src/main/java/io/questdb/PropServerConfiguration.java]: Ok, removed todos from here. Leaved log as is - not sure that queryId will be actually helpful...

File: core/src/main/java/io/questdb/griffin/ExpressionParser.java
Comment (sivukhin) [reply core/src/main/java/io/questdb/griffin/ExpressionParser.java]: Ok, removed todos from here. Leaved log as is - not sure that queryId will be actually helpful...

**Review by puzpuzpuz — APPROVED** [review]
LGTM. Thanks for another solid contribution!

There is a conflict due to a recent test change.

**Review by puzpuzpuz — DISMISSED** [review]

**Comment by puzpuzpuz** [comment]
@sivukhin our build doesn't like your javadocs. :)

Is it ok if I send a PR to fix that against your fork?

**Comment by sivukhin** [comment]
@puzpuzpuz, yes, sure. You can also push directly to my fork (sent you an invite; but feel free to decline and create PR into the fork)

**Review by puzpuzpuz — APPROVED** [review]

**Pull request merged**
